### PR TITLE
RUN-1522: fix and guard asserts related to script invocations

### PR DESCRIPTION
**Pull request by Domiii**
Explanation: https://linear.app/replay/issue/RUN-1488#comment-8f054abf
-> This also includes https://github.com/replayio/chromium-v8/pull/112.

* Add some asserts in `chromium-v8`
* Mark our own code execution paths with `EventsDisallowed`
* Test [here](https://github.com/replayio/build-test-dashboard/blob/master/5276ff1f-a4be-4245-bcd3-a29b4ed7c7d8): 95 passed, 5 failed
  * Crashes [here](https://ui.honeycomb.io/replay/datasets/backend/result/fdndermymRq?tab=raw): 38 `Pause.evaluateInFrame` crashes ❌
* Re-tested [here](https://github.com/replayio/build-test-dashboard/blob/master/bc11dee4-4ca7-43dc-807a-c8e8564d9fd0): 45 passed, 4 failed (50 total)
  * Crashes [here](https://ui.honeycomb.io/replay/datasets/backend/result/x4LUMrNp7Ld): Looks fine. ✔
* Re-tested [here](https://github.com/replayio/build-test-dashboard/blob/master/f10f0340-a3f4-4e98-8120-0c8c3e4a54fa) (after putting debugger behind `IsReplaying` flag): 88 passed, 12 failed
  * Crashes [here](https://ui.honeycomb.io/replay/datasets/backend/result/fbpYd6AuiAs): 23 `ControllerFinished`, 2 `getAllFrames` ✔

### CODE REVIEW

**Comment by Domiii** [comment]
@bhackett1024 After adding these guards, I get a bunch of these command crashes (see [here](https://ui.honeycomb.io/replay/datasets/backend/result/fdndermymRq)). It appears that disallowing events during `SetupRecordReplayCommands` (or maybe in `SendMessageToFrontend`) made things take different paths in such a way that the scope is now affected by it?

What's even weirder is that I do not see this locally after running a bunch of tests. Might be hard to pin down.

Open to any suggestions.

```json
{
  "Chromium": true,
  "Download Speed (Mbps)": 0,
  "Gecko": false,
  "Recordings Prefix": false,
  "Recordings S3 Bucket Prefix": false,
  "Resources Prefix": false,
  "Timestamp": "2023-03-21T11:01:29.198Z",
  "coalesced.buildId": "linux-chromium-20230321-1cd1ac564ffe-bfa354b141e5",
  "controlId": "b9ab764b-5b04-4c56-9f1a-8b0030fbbc5c",
  "enduser.auth_type": "none",
  "event.date": "Tue, 21 Mar 2023",
  "level": "error",
  "message": "TelemetryEvent",
  "recording.isCypress": false,
  "recordingId": "a9233b5f-519c-4951-8233-0d6574407acc",
  "replay.recording.isCypress": false,
  "service_name": "telemetry-server",
  "telemetry.category": "ProtocolMessageCommandFailedCrashReport",
  "telemetry.data.buildId": "linux-chromium-20230321-1cd1ac564ffe-bfa354b141e5",
  "telemetry.data.buildMetadata.date": "20230321",
  "telemetry.data.buildMetadata.platform": "linux",
  "telemetry.data.buildMetadata.runtime": "chromium",
  "telemetry.data.controlId": "b9ab764b-5b04-4c56-9f1a-8b0030fbbc5c",
  "telemetry.data.crashReport.bucket": "",
  "telemetry.data.crashReport.bucketFixed": false,
  "telemetry.data.crashReport.buildId": "linux-chromium-20230321-1cd1ac564ffe-bfa354b141e5",
  "telemetry.data.crashReport.controlId": "b9ab764b-5b04-4c56-9f1a-8b0030fbbc5c",
  "telemetry.data.crashReport.description": "",
  "telemetry.data.crashReport.forTest": true,
  "telemetry.data.crashReport.linkerVersion": "linker-linux-10620-bfa354b141e5",
  "telemetry.data.crashReport.process.backtrace": {
    "frames": [
      "V8_Fatal(char.const*,....)+451",
      "v8::internal::ScopeIterator::HasContext().const+82",
      "v8::internal::DebugEvaluate::ContextBuilder::ContextBuilder(v8::internal::Isolate*,.v8::internal::JavaScriptFrame*,.int)+356",
      "v8::internal::DebugEvaluate::Local(v8::internal::Isolate*,.v8::internal::StackFrameId,.int,.v8::internal::Handle<v8::internal::String>,.bool)+224",
      "v8::internal::DebugStackTraceIterator::Evaluate(v8::Local<v8::String>,.bool)+97",
      "v8_inspector::V8DebuggerAgentImpl::evaluateOnCallFrame(v8_inspector::String16.const&,.v8_inspector::String16.const&,.v8_crdtp::detail::ValueMaybe<v8_inspector::String16>,.v8_crdtp::detail::ValueMaybe<bool>,.v8_crdtp::detail::ValueMaybe<bool>,.v8_crdtp::detail::ValueMaybe<bool>,.v8_crdtp::detail::ValueMaybe<bool>,.v8_crdtp::detail::ValueMaybe<bool>,.v8_crdtp::detail::ValueMaybe<double>,.std::Cr::unique_ptr<v8_inspector::protocol::Runtime::RemoteObject,.std::Cr::default_delete<v8_inspector::protocol::Runtime::RemoteObject>.>*,.v8_crdtp::detail::PtrMaybe<v8_inspector::protocol::Runtime::ExceptionDetails>*)+614",
      "v8_inspector::protocol::Debugger::DomainDispatcherImpl::evaluateOnCallFrame(v8_crdtp::Dispatchable.const&)+476",
      "v8_crdtp::UberDispatcher::DispatchResult::Run()+36",
      "v8_inspector::V8InspectorSessionImpl::dispatchProtocolMessage(v8_inspector::StringView)+474",
      "blink::SendCDPMessage(v8::FunctionCallbackInfo<v8::Value>.const&)+426",
      "v8::internal::FunctionCallbackArguments::Call(v8::internal::CallHandlerInfo)+345",
      "v8::internal::MaybeHandle<v8::internal::Object>.v8::internal::(anonymous.namespace)::HandleApiCallHelper<false>(v8::internal::Isolate*,.v8::internal::Handle<v8::internal::HeapObject>,.v8::internal::Handle<v8::internal::FunctionTemplateInfo>,.v8::internal::Handle<v8::internal::Object>,.unsigned.long*,.int)+1013",
      "v8::internal::Builtin_Impl_HandleApiCall(v8::internal::BuiltinArguments,.v8::internal::Isolate*)+296",
      "0x100a3ff15b38",
      "0x100a3fe8b82c",
      "0x100a3fe8b82c",
      "0x100a3fe8b82c",
      "0x100a3fe8b82c",
      "0x100a3fe89e5c",
      "0x100a3fe89b87",
      "v8::internal::(anonymous.namespace)::Invoke(v8::internal::Isolate*,.v8::internal::(anonymous.namespace)::InvokeParams.const&)+2419",
      "v8::internal::Execution::Call(v8::internal::Isolate*,.v8::internal::Handle<v8::internal::Object>,.v8::internal::Handle<v8::internal::Object>,.int,.v8::internal::Handle<v8::internal::Object>*)+284",
      "v8::internal::CommandCallback(char.const*,.char.const*)+658",
      "recordreplay::SendCommand(char.const*,.Json::Value.const&,.bool)+426",
      "EvaluateInFrameCommand::Process(Json::Value.const&,.std::__cxx11::basic_string<char,.std::char_traits<char>,.std::allocator<char>.>&)+132",
      "CommandRequest::Process(Json::Value.const&)+516",
      "recordreplay::PerformPauseDataRequest(Json::Value.const&)+276",
      "PauseRequestHandler::Start(Json::Value.const&)+210",
      "ManifestFinished(Json::Value*)+751",
      "RunToPointHandler::ReachedDestination(Json::Value.const&)+491",
      "recordreplay::OnScanTargetHit()+36",
      "recordreplay::OnInstrument(char.const*,.char.const*,.int)+1523",
      "v8::internal::OnInstrumentation(v8::internal::Isolate*,.v8::internal::Handle<v8::internal::JSFunction>,.int)+208",
      "v8::internal::Runtime_RecordReplayInstrumentation(int,.unsigned.long*,.v8::internal::Isolate*)+270",
      "0x100a3ff15a38",
      "0x100a3fffad11",
      "0x100a3fe8b82c",
      "0x100a3fe8b82c",
      "0x100a3fe8b82c",
      "0x100a3fe89e5c",
      "0x100a3fe89b87",
      "v8::internal::(anonymous.namespace)::Invoke(v8::internal::Isolate*,.v8::internal::(anonymous.namespace)::InvokeParams.const&)+2419",
      "v8::internal::Execution::Call(v8::internal::Isolate*,.v8::internal::Handle<v8::internal::Object>,.v8::internal::Handle<v8::internal::Object>,.int,.v8::internal::Handle<v8::internal::Object>*)+284",
      "v8::Function::Call(v8::Local<v8::Context>,.v8::Local<v8::Value>,.int,.v8::Local<v8::Value>*)+648",
      "blink::V8ScriptRunner::CallFunction(v8::Local<v8::Function>,.blink::ExecutionContext*,.v8::Local<v8::Value>,.int,.v8::Local<v8::Value>*,.v8::Isolate*)+915",
      "blink::bindings::CallbackInvokeHelper<blink::CallbackFunctionBase,.(blink::bindings::CallbackInvokeHelperMode)2,.(blink::bindings::CallbackReturnTypeIsPromise)0>::Call(int,.v8::Local<v8::Value>*)+91",
      "blink::V8EventHandlerNonNull::InvokeWithoutRunnabilityCheck(blink::bindings::V8ValueOrScriptWrappableAdapter,.blink::HeapVector<blink::ScriptValue,.0u>.const&)+365",
      "blink::JSEventHandler::InvokeInternal(blink::EventTarget&,.blink::Event&,.v8::Local<v8::Value>)+1326",
      "blink::JSBasedEventListener::Invoke(blink::ExecutionContext*,.blink::Event*)+744",
      "blink::EventTarget::FireEventListeners(blink::Event&,.blink::EventTargetData*,.blink::HeapVector<blink::RegisteredEventListener,.1u>&)+1851",
      "blink::EventTarget::FireEventListeners(blink::Event&)+361",
      "blink::EventTarget::DispatchEventInternal(blink::Event&)+64",
      "blink::XMLHttpRequest::DispatchReadyStateChangeEvent()+680",
      "blink::XMLHttpRequest::EndLoading()+205",
      "blink::XMLHttpRequest::DidFinishLoadingInternal()+328",
      "blink::XMLHttpRequest::DidFinishLoading(unsigned.long)+223",
      "blink::ThreadableLoader::NotifyFinished(blink::Resource*)+266",
      "blink::Resource::NotifyFinished()+231",
      "blink::ResourceFetcher::HandleLoaderFinish(blink::Resource*,.base::TimeTicks,.blink::ResourceFetcher::LoaderFinishType,.unsigned.int,.bool)+769",
      "blink::ResourceLoader::DidFinishLoading(base::TimeTicks,.long,.long,.long,.bool,.absl::optional<bool>)+839",
      "non-virtual.thunk.to.blink::ResourceLoader::DidFinishLoadingBody()+96",
      "blink::ResponseBodyLoader::OnStateChange()+1311",
      "blink::ResourceLoader::DidFinishLoading(base::TimeTicks,.long,.long,.long,.bool,.absl::optional<bool>)+934",
      "blink::WebURLLoader::Context::OnCompletedRequest(network::URLLoaderCompletionStatus.const&)+362",
      "blink::WebResourceRequestSender::OnRequestComplete(network::URLLoaderCompletionStatus.const&)+763",
      "blink::MojoURLLoaderClient::FlushDeferredMessages()+475",
      "blink::MojoURLLoaderClient::StoreAndDispatch(std::Cr::unique_ptr<blink::MojoURLLoaderClient::DeferredMessage,.std::Cr::default_delete<blink::MojoURLLoaderClient::DeferredMessage>.>)+173",
      "blink::ThrottlingURLLoader::OnComplete(network::URLLoaderCompletionStatus.const&)+504",
      "network::mojom::URLLoaderClientStubDispatch::Accept(network::mojom::URLLoaderClient*,.mojo::Message*)+605",
      "mojo::InterfaceEndpointClient::HandleValidatedMessage(mojo::Message*)+753",
      "mojo::MessageDispatcher::Accept(mojo::Message*)+211",
      "mojo::InterfaceEndpointClient::HandleIncomingMessage(mojo::Message*)+58",
      "mojo::internal::MultiplexRouter::ProcessIncomingMessage(mojo::internal::MultiplexRouter::MessageWrapper*,.mojo::internal::MultiplexRouter::ClientCallBehavior,.base::SequencedTaskRunner*)+1022",
      "mojo::internal::MultiplexRouter::Accept(mojo::Message*)+363",
      "mojo::MessageDispatcher::Accept(mojo::Message*)+211",
      "mojo::Connector::DispatchMessage(mojo::ScopedHandleBase<mojo::MessageHandle>)+471",
      "mojo::Connector::ReadAllAvailableMessages()+244",
      "base::RepeatingCallback<void.(int)>::Run(int).const.&+39",
      "mojo::SimpleWatcher::OnHandleReady(int,.unsigned.int,.mojo::HandleSignalsState.const&)+291",
      "base::internal::Invoker<base::internal::BindState<void.(mojo::SimpleWatcher::*)(int,.unsigned.int,.mojo::HandleSignalsState.const&),.base::WeakPtr<mojo::SimpleWatcher>,.int,.unsigned.int,.mojo::HandleSignalsState>,.void.()>::RunOnce(base::internal::BindStateBase*)+142",
      "base::TaskAnnotator::RunTaskImpl(base::PendingTask&)+257",
      "base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoWorkImpl(base::LazyNow*)+965",
      "base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoWork()+139",
      "non-virtual.thunk.to.base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoWork()+21",
      "base::MessagePumpDefault::Run(base::MessagePump::Delegate*)+154",
      "base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::Run(bool,.base::TimeDelta)+355",
      "base::RunLoop::Run(base::Location.const&)+461",
      "content::RendererMain(content::MainFunctionParams)+1434",
      "content::RunOtherNamedProcessTypeMain(std::Cr::basic_string<char,.std::Cr::char_traits<char>,.std::Cr::allocator<char>.>.const&,.content::MainFunctionParams,.content::ContentMainDelegate*)+686",
      "content::ContentMainRunnerImpl::Run()+637",
      "content::RunContentProcess(content::ContentMainParams,.content::ContentMainRunner*)+294",
      "content::ContentMain(content::ContentMainParams)+95",
      "ChromeMain+525",
      "new_libc_start_main(void.(*)(int,.char.const**))+22",
      "_start+42",
      "onstack:V8_Fatal(char.const*,....)+451",
      "onstack:v8::internal::Object::TransitionAndWriteDataProperty(v8::internal::LookupIterator*,.v8::internal::Handle<v8::internal::Object>,.v8::internal::PropertyAttributes,.v8::Maybe<v8::internal::ShouldThrow>,.v8::internal::StoreOrigin)+303",
      "onstack:v8::internal::Object::SetPropertyOrElement(v8::internal::Isolate*,.v8::internal::Handle<v8::internal::Object>,.v8::internal::Handle<v8::internal::Name>,.v8::internal::Handle<v8::internal::Object>,.v8::Maybe<v8::internal::ShouldThrow>,.v8::internal::StoreOrigin)+309",
      "onstack:bool.std::Cr::__function::__policy_invoker<bool.(v8::internal::Handle<v8::internal::String>,.v8::internal::Handle<v8::internal::Object>,.v8::internal::ScopeIterator::ScopeType)>::__call_impl<std::Cr::__function::__default_alloc_func<v8::internal::ScopeIterator::ScopeObject(v8::internal::ScopeIterator::Mode)::$_1,.bool.(v8::internal::Handle<v8::internal::String>,.v8::internal::Handle<v8::internal::Object>,.v8::internal::ScopeIterator::ScopeType)>.>(std::Cr::__function::__policy_storage.const*,.v8::internal::Handle<v8::internal::String>&&,.v8::internal::Handle<v8::internal::Object>&&,.v8::internal::ScopeIterator::ScopeType)+194",
      "onstack:v8::internal::ScopeIterator::VisitLocals(std::Cr::function<bool.(v8::internal::Handle<v8::internal::String>,.v8::internal::Handle<v8::internal::Object>,.v8::internal::ScopeIterator::ScopeType)>.const&,.v8::internal::ScopeIterator::Mode,.v8::internal::ScopeIterator::ScopeType).const+468",
      "onstack:v8::internal::FrameInspector::GetExpression(int)+61",
      "onstack:v8::internal::ScopeIterator::VisitLocalScope(std::Cr::function<bool.(v8::internal::Handle<v8::internal::String>,.v8::internal::Handle<v8::internal::Object>,.v8::internal::ScopeIterator::ScopeType)>.const&,.v8::internal::ScopeIterator::Mode,.v8::internal::ScopeIterator::ScopeType).const+236",
      "onstack:v8::internal::FactoryBase<v8::internal::Factory>::NewFixedArrayWithFiller(v8::internal::Handle<v8::internal::Map>,.int,.v8::internal::Handle<v8::internal::Oddball>,.v8::internal::AllocationType)+68",
      "onstack:chrome+22405751",
      "onstack:v8::internal::Factory::NewSlowJSObjectFromMap(v8::internal::Handle<v8::internal::Map>,.int,.v8::internal::AllocationType,.v8::internal::Handle<v8::internal::AllocationSite>)+83",
      "onstack:v8::internal::ScopeIterator::HasContext().const+82",
      "onstack:v8::internal::DebugEvaluate::ContextBuilder::ContextBuilder(v8::internal::Isolate*,.v8::internal::JavaScriptFrame*,.int)+356",
      "onstack:v8::internal::DebugEvaluate::Local(v8::internal::Isolate*,.v8::internal::StackFrameId,.int,.v8::internal::Handle<v8::internal::String>,.bool)+224",
      "onstack:new_pthread_mutex_lock(void*)+21",
      "onstack:chrome+222086992",
      "onstack:chrome+222087112",
      "onstack:chrome+222087232",
      "onstack:chrome+222088568",
      "onstack:chrome+222089584",
      "onstack:chrome+222089704",
      "onstack:chrome+222088808",
      "onstack:chrome+222088928",
      "onstack:chrome+222088688",
      "onstack:InstructionPoint"
    ],
    "threadId": 1
  },
  "telemetry.data.crashReport.process.manifest": {
    "kind": "pauseRequest",
    "requestMethod": "Pause.evaluateInFrame",
    "requestType": "command"
  },
  "telemetry.data.crashReport.process.text": "Check failed: needs_context && current_scope_ == closure_scope_ && current_scope_->is_function_scope() && !function_->is_null() implies function_->context() != *context_.",
  "telemetry.data.crashReport.process.threadId": 1,
  "telemetry.data.crashReport.process.why": "LinkerApplicationCrash",
  "telemetry.data.crashReport.processId": 108,
  "telemetry.data.crashReport.processNonFatal.threadId": 0,
  "telemetry.data.crashReport.processNonFatal.why": "MissingReport",
  "telemetry.data.crashReport.recordingAssertionsDisabled": false,
  "telemetry.data.crashReport.recordingFeaturesDisabled": "",
  "telemetry.data.crashReport.recordingId": "a9233b5f-519c-4951-8233-0d6574407acc",
  "telemetry.data.crashReport.sessionIds": [
    "b9ab764b-5b04-4c56-9f1a-8b0030fbbc5c/7dd15050-90b1-4cb7-9b3d-897d3d1b2326"
  ],
  "telemetry.data.crashReport.why": "NonFatal",
  "telemetry.data.crashReportUuid": "1fd4de0a-6ef6-4afb-9c43-073ee23985cd",
  "telemetry.data.event": "undefined",
  "telemetry.data.forTest": true,
  "telemetry.data.protocolMessageId": 267,
  "telemetry.data.recording.internal": true,
  "telemetry.data.recording.isCypress": false,
  "telemetry.data.recording.userEmail": "internal-tests@replay.io",
  "telemetry.data.recording.userPriority": 5,
  "telemetry.data.recording.workspaceName": "Replay Internal Livetests",
  "telemetry.data.recordingId": "a9233b5f-519c-4951-8233-0d6574407acc",
  "telemetry.data.requestCommandMethod": "Pause.evaluateInFrame",
  "telemetry.data.requestOptions.shortTimeout": true,
  "telemetry.data.requestType": "command",
  "telemetry.data.service_name": "controller",
  "telemetry.data.sessionId": "b9ab764b-5b04-4c56-9f1a-8b0030fbbc5c/7dd15050-90b1-4cb7-9b3d-897d3d1b2326",
  "telemetry.data.success": "n/a",
  "telemetry.data.timestamp": 1679396489198,
  "telemetry.userData": "undefined",
  "userEmail": "internal-tests@replay.io"
}
```

File: third_party/blink/renderer/bindings/core/v8/record_replay_interface.cc [hunk 1]
```diff
@@ -3062,6 +3062,8 @@ static void SetCDPMessageCallback(const v8::FunctionCallbackInfo<v8::Value>& arg
 }
 
 static void SendMessageToFrontend(const v8_inspector::StringView& message) {
+  recordreplay::AutoDisallowEvents disallow(
+      "RecordReplay_SendMessageToFrontend");
```
Comment (bhackett1024) [inline third_party/blink/renderer/bindings/core/v8/record_replay_interface.cc:3066]: Can we check IsReplaying() here?

File: third_party/blink/renderer/bindings/core/v8/record_replay_interface.cc [hunk 2]
```diff
@@ -3107,6 +3109,7 @@ RecordReplayRegisterV8Inspector(v8_inspector::V8Inspector* inspector,
     // For now we only connect to the first frame.
     static int ContextGroupId = 1;
 
+    recordreplay::AutoDisallowEvents disallow("RecordReplayRegisterV8Inspector");
```
Comment (bhackett1024) [inline third_party/blink/renderer/bindings/core/v8/record_replay_interface.cc:3112]: Can we add IsReplaying() to the test above this?

**Review by bhackett1024 — APPROVED** [review]
The PR looks fine, I don't know why the command crashes are happening and I'm not really sure what you're asking for.  Looking at that stack trace there is V8_Fatal on the stack so V8 is crashing intentionally for some reason and I'd look into why.  From the stack this is also happening during an evaluateInFrame command, have you tried manually testing that?

**Comment by Domiii** [comment]
@bhackett1024 
> The PR looks fine, I don't know why the command crashes are happening and I'm not really sure what you're asking for. Looking at that stack trace there is V8_Fatal on the stack so V8 is crashing intentionally for some reason and I'd look into why. From the stack this is also happening during an evaluateInFrame command, have you tried manually testing that?

That `evaluateInFrame` is part of the initial `runToPoint`, and I did test this locally, and I re-tested it online, and it did not show up.

The weird part is that in that one test run it happened so consistently, and now its just gone, despiste both runs having the same `buildId`. Very odd.

**Comment by Domiii** [comment]
@bhackett1024 Yeah, I'd love to know who is sending it. I have looked into this in the past for a little while, but have not found out yet who sends the commands.

File: third_party/blink/renderer/bindings/core/v8/record_replay_interface.cc (hunk 2 above)
Comment (Domiii) [reply third_party/blink/renderer/bindings/core/v8/record_replay_interface.cc:3112]: I guess you mentioned in the past, that we did not want the debugger active during recording?

I'll add it and run a few tests. Fingers crossed, it won't cause a lot of new divergences.

Comment (Domiii) [reply third_party/blink/renderer/bindings/core/v8/record_replay_interface.cc:3115]: As expected, there is some more follow-up coming out of this: https://linear.app/replay/issue/RUN-1596/mismatch-in-v8consolemessagestorageaddmessage-due-to-replay-only

**Pull request merged**